File: RDMPEG/RDMPEGPlayer/RDMPEGRenderScheduler/RDMPEGRenderScheduler.m
//
//  RDMPEGRenderScheduler.m
//  RDMPEG
//
//  Created by Serhii Alpieiev on 9/8/17.
//  Copyright © 2017 Readdle. All rights reserved.
//

#import "RDMPEGRenderScheduler.h"
#import <Log4Cocoa/Log4Cocoa.h>

#import <RDMPEG/RDMPEG-Swift.h>


NS_ASSUME_NONNULL_BEGIN

@interface RDMPEGRenderScheduler ()

@property (nonatomic, strong, nullable) NSTimer *timer;
@property (nonatomic, strong, nullable) RDMPEGRenderSchedulerCallback callback;

@end



@implementation RDMPEGRenderScheduler

#pragma mark - Overridden Class Methods

+ (L4Logger *)l4Logger {
    return [L4Logger loggerForName:@"rd.mediaplayer.RDMPEGRenderScheduler"];
}

#pragma mark - Lifecycle

- (void)dealloc {
    [self stop];
}

#pragma mark - Public Accessors

- (BOOL)isScheduling {
    return self.timer != nil;
}

#pragma mark - Public Methods

- (void)startWithCallback:(RDMPEGRenderSchedulerCallback)callback {
    if (self.isScheduling) {
        log4Assert(NO, @"Already scheduling");
        return;
    }
    
    self.callback = callback;
    
    
    RDMPEGWeakTimerTarget *timerTarget = [[RDMPEGWeakTimerTarget alloc] initWithTarget:self action:@selector(renderTimerFired:)];
    self.timer = [NSTimer timerWithTimeInterval:0.0 target:timerTarget selector:@selector(timerFired:) userInfo:nil repeats:YES];
    [[NSRunLoop mainRunLoop] addTimer:self.timer forMode:NSRunLoopCommonModes];
}

- (void)stop {
    if (self.isScheduling == NO) {
        return;
    }
    
    [self.timer invalidate];
    self.timer = nil;
    
    self.callback = nil;
}

#pragma mark - Timer

- (void)renderTimerFired:(NSTimer *)timer {
    @autoreleasepool {
        NSDate *nextFireDate = self.callback();
        if (nextFireDate == nil) {
            nextFireDate = [NSDate dateWithTimeIntervalSinceNow:0.01];
        }
        timer.fireDate = nextFireDate;
    }
}

@end

NS_ASSUME_NONNULL_END

File: RDMPEG/RDMPEGPlayer/RDMPEGRenderScheduler/RDMPEGRenderScheduler.swift
//
//  RDMPEGRenderScheduler.swift
//  RDMPEG
//
//  Created by Max Berezhnoy on 15/07/2024.
//  Copyright © 2024 Readdle. All rights reserved.
//

import Foundation
import Log4Cocoa

class RDMPEGRenderScheduler: NSObject {
    private var timer: Timer?
    private var callback: (() -> Date?)?

    var isScheduling: Bool {
        return timer != nil
    }

    deinit {
        stop()
    }

    func start(with callback: @escaping () -> Date?) {
        guard isScheduling == false else {
            log4Assert(false, "Already scheduling")
            return
        }

        self.callback = callback

        let newTimer = Timer(timeInterval: 0.0, repeats: true) { [weak self] in
            self?.renderTimerFired($0)
        }

        RunLoop.main.add(newTimer, forMode: .common)

        timer = newTimer
    }

    func stop() {
        guard isScheduling else { return }

        timer?.invalidate()
        timer = nil
        callback = nil
    }

    private func renderTimerFired(_ timer: Timer) {
        let nextFireDate = callback?() ?? Date(timeIntervalSinceNow: 0.01)
        timer.fireDate = nextFireDate
    }
}

extension RDMPEGRenderScheduler {
    class var l4Logger: L4Logger {
        return L4Logger(forName: "rd.mediaplayer.RDMPEGRenderScheduler")
    }
}
